File: PRO2/2020/P2/center_list.c
/*
* TITLE: PROGRAMMING II LABS
* SUBTITLE: Practical 2
* AUTHOR 1: Inés Faro Pérez LOGIN 1: ines.fperez@udc.es
* AUTHOR 2: Andrea Lama Muiño LOGIN 2: a.lama1@udc.es
* GROUP: 4.4
* DATE: 8/05/20
*/

#include "types.h"
#include "party_list.h"
#include "center_list.h"

void createEmptyListC (tListC *L){
    L->index = NULLC;
}

bool isEmptyListC (tListC L){
    return L.index == NULLC;
}

tPosC firstC (tListC L){
    return 0;
}

tPosC lastC (tListC L){
    return L.index;
}

tPosC nextC (tPosC p, tListC L){
    if (p == lastC(L)) {
        return NULLC;
    } else {
        return ++p;
    }
}

tPosC previousC (tPosC p, tListC L){
    return --p;
}
bool insertItemC(tItemC d, tListC *C){
    tPosC i;

    if (C -> index == MAX-1){ //comprobamos si el array está lleno
        return false;
    } else{
        if (isEmptyListC(*C) || strcmp(d.centerName,(*C).data[C->index].centerName)> 0) {
            C->index++;
            C->data[C->index] = d;
        }else{
            C -> index++;
            for(i=C->index; (i>0) && strcmp(d.centerName,C->data[i-1].centerName)<0;i--)
                C -> data[i] = C -> data[i-1];
            C -> data[i] = d;
        }
        return true;
    }
}
void updateListC (tList A, tPosC p, tListC *L){
    L->data[p].partyList = A; //Reescribimos la lista de partidos
}

void deleteAtPositionC (tPosC p, tListC *L){
    tPosC i;
    L->index --;
    for(i=p; i<=L->index; i++)
        L->data[i] = L->data[i+1];

}

tItemC getItemC (tPosC p, tListC L){
    return L.data[p];
}

void updateValidVotesC (tNumVotes num, tPosC p, tListC *L){
    L->data[p].validVotes = num;
}

void updateNullVotesC (tNumVotes num, tPosC p, tListC *L){
    L->data[p].nullVotes = num;
}

tPosC findItemC(tCenterName d, tListC L){
    tPosC p;

    if(L.index == NULLC)
        return NULLC;
    else if ((strcmp(d, L.data[0].centerName)<0)||(strcmp(d, L.data[L.index].centerName)>0))
        return NULLC;
    else{
        for(p = 0; (p<L.index)&& strcmp(L.data[p].centerName,d)<0;p++);
        if(strcmp(L.data[p].centerName,d)==0)
            return p;
        else
            return NULLC;
    }
}

File: PRO2/1ºvez/entregapro/CenterList.pas
{program RecuentoVotos;}
{TITLE: PROGRAMMING II LABS
* SUBTITLE: Practical 2
* AUTHOR 1: Javier López Durán LOGIN 1: javier.lduran
* AUTHOR 2: Inés Faro Pérez LOGIN 2: ines.fperez
* GROUP: 3.1
* DATE: 03/05/19}

unit CenterList;
interface
 uses SharedTypes, PartyList;	
 const
	NULLC = 0;
	beg = 1;
	MAX = 25;
	
 type	
 	tNumVotes = integer; 	
 	tPartyName = string; 		
	tItemC = record 
		centername: tCenterName;
		totalvoters: tNumVotes;
		validvotes: tNumVotes;
		partyList: tList;
		end;
	tPosC = NULLC .. MAX; 
	tListC = record 
		dato: array[beg..MAX] of tItemC;
		index: tPosC;		
		end;				
		
	procedure createEmptyListC(var L: tListC);
	function isEmptyListC(L: tListC): boolean;
	function firstC(L: tListC): tPosC;
	function lastC(L: tListC): tPosC;
	function nextC(p: tPosC; L: tListC): tPosC;
	function previousC(p: tPosC; L: tListC): tPosC;
	function insertItemC(I: tItemC; var L: tListC): boolean;
	procedure deleteAtPositionC(p: tPosC; var L: tListC);
	function getItemC(p: tPosC; L: tListC): tItemC;
	function updateValidVotersC(nv: tNumVotes; p: tPosC; var L: tListC):tListC;
	procedure updateListC(L: tList; p: tPosC; var LC: tListC);
	function findItemC(cn: tCenterName; L: tListC):tPosC;

implementation 

	procedure createEmptyListC(var L: tListC);
	begin	
		L.index := NULLC;		
	end;
				
(**********************************************************************)
	function isEmptyListC(L: tListC): boolean;
	begin
		isEmptyListC := (L.index = NULLC);
	end;
		
(**********************************************************************)

	function firstC(L: tListC): tPosC;
	begin	
		firstC := beg;				
	end;
		
(**********************************************************************)

	function lastC(L: tListC): tPosC;
	begin
		lastC := L.index;
	end;

(**********************************************************************)

	function nextC(p: tPosC; L: tListC): tPosC;
	begin
		if p = L.index then
			nextC := NULLC
		else
			nextC := p + 1;	
		
	end;
		
(**********************************************************************)

	function previousC(p: tPosC; L: tListC): tPosC;	
	begin
		if p = firstC(L) then
			previousC := NULLC
		else
			previousC := p - 1;	
		
	end;
		
(**********************************************************************)

	function insertItemC(I: tItemC; var L: tListC): boolean;
	var
		w : tPosC;
	begin
		if L.index = MAX then 
			insertItemC := false
		else begin
			if isEmptyListC(L) then begin 
				insertItemC := true;
				L.index:= L.index + 1; 
				L.dato[beg] := I;
			end else begin
				insertItemC := true;
				L.index:= L.index + 1; 
				w := L.index;
				L.dato[w] := I;
				while L.dato[w-1].centername > I.centername do begin
					L.dato[w] := L.dato[w-1];
					w := w - 1;
				end;
				L.dato[w] := I;
			end;	
		end;
	
end;
	
(**********************************************************************)

	procedure deleteAtPositionC(p: tPosC; var L: tListC);
	var
		i: tPosC;
	begin
		L.index := L.index - 1; 
		for i := p to L.index do
			L.dato[i] := L.dato[i+1];
	end;	

(**********************************************************************)

	function getItemC(p: tPosC; L: tListC): tItemC;
	begin
		getItemC := L.dato[p];
	end;
		
(**********************************************************************)
	
	function updateValidVotersC(nv: tNumVotes; p: tPosC; var L: tListC):tListC;
	begin	
		L.dato[p].validvotes := nv;
		updateValidVotersC:= L;
	end;

(**********************************************************************)

	procedure updateListC(L: tList; p: tPosC; var LC: tListC);
	begin
		LC.dato[p].partylist := L;	
	end;
		
(**********************************************************************)
 
	function findItemC(cn: tCenterName; L: tListC):tPosC;
	var	
		p: tPosC;
			
	begin
			p := firstC(L);
			while (p < L.index) and (L.dato[p].centername <> cn) do
				p := nextC(p,L);
			if (L.dato[p].centername = cn) then
				findItemC := p
			else
				findItemC := NULLC;
	end;		

			
END.



